File: leetcode/java/linkedlist/2.java
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode root = new ListNode();
        ListNode currentNode = root;
        int carry = 0;
        while (l1 != null || l2 != null){
            int val1 = l1 != null ? l1.val : 0;
            int val2 = l2 != null ? l2.val : 0;
            int sum = val1 + val2 + carry;
            carry = sum / 10;
            currentNode.next = new ListNode(sum % 10);
            currentNode = currentNode.next;
            l1 = l1 != null ? l1.next : null;
            l2 = l2 != null ? l2.next : null;
        }
        if (carry == 1){
            currentNode.next = new ListNode(carry);
        }
        return root.next;
    }
}

File: leetcode/cpp/linkedlist/2.cpp
//Leetcode Problem #2 : Add Two Numbers
//Link to problem : https://leetcode.com/problems/add-two-numbers/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * }; 
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode * head = new ListNode(-1);   
        ListNode * tail = head;
        int carry = 0; //to calculate carry while adding digits
        while (l1 || l2 || carry) {
            int n1 = l1 ? l1->val : 0; //Adding zero doesn't change sum
            int n2 = l2 ? l2->val : 0;
            int sum = n1 + n2 + carry;
            carry = sum / 10;
            tail->next = new ListNode(sum % 10);
            tail = tail->next;
            l1 ? l1 = l1->next : NULL;
            l2 ? l2 = l2->next : NULL;
        }
        if (carry) 
            tail->next = new ListNode(1);
        
        return head->next;  
    }
};